Feature/git versioning

galiron opened the pull request:
## :loudspeaker: Type of change
<!--
Please delete options that are not relevant.
-->

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## :scroll: Description
<!--
Please include a short summary of the change.
If it is a new feature tell is why we would need this.
-->

## :link: Related Issues
* Fixes # (issue)

## :ballot_box_with_check: Checklist:

- [x] I have performed a self-review of my own code
- [x] I have run the auto code formatting scripts <!-- npm run fix, gradle spotlessApply -->
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I updated the `CHANGELOG.md`
- [x] I have added tests that prove my fix is effective or that my feature works

galiron committed:
Remove try/catch to throw error

galiron committed:
Versionierungs Menü zum Online IDE Reiter hinzugefügt

Das Versionierungs Menü wurde der Online IDE und
dessen Reiter hinzugefügt, um für eine bessere
Usability zu sorgen. Außerdem wurden die props:
commitMessage und versionId aus der VersionSaveComponent.tsx
entfernt, da sie nicht länger benötigt wurden.

HannesOlszewski commented on client/src/components/IdeIframe.tsx:
```diff
@@ -1,18 +1,90 @@
+        messageService.success('Version got changed')
+      } else {
+        messageService.error('Selected version is the current version')
+      }
+      result.refetch()
+    },
+    refetchQueries: [
+      {
+        query: ListFilesDocument,
+        variables: {
+          context: {
+            id,
+            type: FileContextType.Answer
+          },
+          path: '/'
+        }
+      }
+    ]
+  })
+
+  const menuItemList = result.data?.answer.versions.map(commit => (
+    <Menu.Item
+      key={commit.versionKey}
+      onClick={() =>
+        changeVersion({
+          variables: { id, versionID: commit.versionKey }
+        }).then(() => result.refetch())
+      }
+    >
+      {commit.commitMessage}
+    </Menu.Item>
+  ))
+
+  const menu = <Menu>{menuItemList}</Menu>
+
+  const dropdown = (
+    <Dropdown overlay={menu}>
+      <Button>{result.data?.answer.currentVersionName}</Button>
+    </Dropdown>
+  )
```
You can refactor this into it's own component and use it in the answer page too

HannesOlszewski commented on client/src/components/IdeIframe.tsx:
```diff
@@ -1,18 +1,90 @@
-      src={ideUrl}
-      title="Edit answer"
-    />
+    <Row gutter={[24, 28]}>
```
Why the horizontal spacing? The two columns are one row each

galiron committed:
code refactoring

refactored code, so that the VersionSaveComponent.tsx now
 contains the whole functionality and is no longer split
 over multiple components. Also adjusted some comments

HannesOlszewski commented on client/src/components/VersionSaveComponent.tsx:
```diff
@@ -1,13 +1,66 @@
+      onClick={() =>
+        changeVersion({
+          variables: { id: props.answerId, versionID: commit.versionKey }
+        }).then(() => result.refetch())
```
Since we already use `refetchQueries` for the Mutation wenn can refetch this query there as well

HannesOlszewski commented on client/src/components/VersionSaveComponent.tsx:
```diff
@@ -1,13 +1,66 @@
+
+  const dropdown = (
+    <Dropdown overlay={menu}>
+      <Button>{result.data?.answer.currentVersionName}</Button>
```
What if `result.data === undefined`?

HannesOlszewski commented on client/src/components/VersionSaveComponent.tsx:
```diff
@@ -33,33 +86,39 @@ const VersionSaveComponent: React.FC<{
-      </Col>
+    <Row justify="end" align="middle" gutter={24}>
+      <Col>Submission version:</Col>
+      <Col style={{ minWidth: '20px' }}>{dropdown}</Col>
```
Is the min width necessary for a column? Will this ever be met?

HannesOlszewski commented on src/main/kotlin/org/codefreak/codefreak/service/file/IFileCollectionRepository.kt:
```diff
@@ -2,17 +2,22 @@ package org.codefreak.codefreak.service.file
 interface IFileCollectionRepository {
   /**
-   * Function to save a version to the filesystem within the backend.
+   * Saving a version to the filesystem within the backend.
```
Saves

galiron committed:
code review fixes

-removed style
-added handler for undefined label result.data on a button
-changed comment

galiron committed:
disabling version feature during evaluation

switching versions while the evaluation is running caused errors
